File: js/치킨 배달(15686).js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, m] = input[0].split(' ').map(Number);
const graph = [];
for (let i=1; i<=n; i++) {
    graph.push(input[i].split(' ').map(Number));
}

const chickenArr = [];
const houseArr = [];

for (let i=0; i<n; i++) {
    for (let j=0; j<n; j++) {
        if (graph[i][j] === 1) {
            houseArr.push([i, j]);
        }
        else if (graph[i][j] === 2) {
            chickenArr.push([i, j]);
        }
    }
}

const arr = [];
let INF = Number.MAX_SAFE_INTEGER;

function backTracking(num, cnt) {
    if (num > chickenArr.length) return;

    if (cnt === m) {
        let resultVal = 0;
        for (let elem of houseArr) {
            let [hx, hy] = elem;
            let MaxVal = Number.MAX_SAFE_INTEGER;

            for (let i=0; i<arr.length; i++) {
                let [cx, cy] = chickenArr[arr[i]];
                MaxVal = Math.min(MaxVal, Math.abs(hx-cx) + Math.abs(hy-cy));
            }
            resultVal += MaxVal
        }
        INF = Math.min(INF, resultVal);
        return;
    }

    arr.push(num);
    backTracking(num+1, cnt+1);
    arr.pop();
    backTracking(num+1, cnt);
    return INF;
}

backTracking(0,0);
console.log(INF);

File: python/치킨 배달(15686).py
N, M = map(int, input().split())
graph = []
house = []
chicken = []

for col in range(N):
    graph.append(list(map(int, input().split())))
    for row in range(N):
        if graph[col][row] == 1:
            house.append((col, row))
        elif graph[col][row] == 2:
            chicken.append((col, row))

arr = []
real_check = int(1e9)

def back_traking(num, cnt):
    global real_check

    if num > len(chicken):
        return

    if cnt == M:
        result_val = 0
        for hx, hy in house:
            min_check = int(1e9)

            for idx in arr:
                cx, cy = chicken[idx]
                min_check = min(min_check, abs(hx-cx) + abs(hy-cy))
            
            result_val += min_check
        
        real_check = min(real_check, result_val)
        return
    
    arr.append(num)
    back_traking(num+1, cnt+1)
    arr.pop()
    back_traking(num+1, cnt)
    return real_check

back_traking(0,0)
print(real_check)